### ISSUE TITLE
Package Proposal: QUIC and HTTP3 plugin

### ISSUE THREAD

**Issue by iapicca** [fluttercommunity/community#59]
Given that
- [http/3 draft](https://tools.ietf.org/html/draft-ietf-quic-http-31) has been published last month
- chrome is [deploying it](https://blog.chromium.org/2020/10/chrome-is-deploying-http3-and-ietf-quic.html)
- on safari is even enabled by default ([source](https://en.wikipedia.org/wiki/HTTP/3))

[HTTP3](https://en.wikipedia.org/wiki/HTTP/3) [QUIC](https://en.wikipedia.org/wiki/QUIC) are very [promising](https://blog.cloudflare.com/last-call-for-quic/) and 
react native approach seems to be through native clients ([source](https://medium.com/the-react-native-log/using-cronet-in-your-mobile-app-7dda3a89c132
)), 
but Flutter Team has, legitimately, [too much on its plate](https://github.com/dart-lang/sdk/issues/38595#issuecomment-712857906) to take craft a dart package [references [1](https://github.com/grpc/grpc-dart/issues/374), [2](https://github.com/dart-lang/sdk/issues/22657), [3](https://github.com/dart-lang/sdk/issues/38595)]

It would be great if the community could come out with a [plugin](https://flutter.dev/docs/development/packages-and-plugins/developing-packages#federated-plugins) to support this feature, 
I don't have the brain muscles or the expertise to deal with it myself, but I'd be happy to help.

I believe there are 2 possible strategies to support `QUIC` :

1. CRONET on the native side for [android](https://developer.android.com/guide/topics/connectivity/cronet) and [ios](https://chromium.googlesource.com/chromium/src/+/master/components/cronet/ios/Cronet.h)

2. [QUICHE](https://github.com/cloudflare/quiche) + [dart:ffi](https://dart.dev/guides/libraries/c-interop) (resource for [rust + dart:ffi](https://medium.com/flutter-community/finally-running-rust-natively-on-a-flutter-plugin-here-is-how-6f2826eb1735)) could work from both android, ios, desktop and maybe in the future for [web](https://github.com/dart-lang/sdk/issues/37882) 

there will be probably some js package to use for web, but I haven't located it 

Do you think such plugin could be useful for the community?

**Comment by jayjah** [comment]
@iapicca 
What do you think about [QUICK](https://github.com/pion/quic)? I know, that the way around with golang may quite not be a good solution at all, cause of this [article](https://sites.google.com/a/athaydes.com/renato-athaydes/posts/buildingamobilefrontendforagoapplicationusingflutter). But I think it is a possible solution at all, at least on native side ...

**Comment by iapicca** [comment]
@jayjah 
I'm not sure the article is still reliable (2018) anyhow I believe c/c++ and rust are an easier for `dart:ffi`
that's the only reason why I mentioned specifically `quiche`
but if you are able to run `quick` that would be awesome, I have no clue how to do that

**Comment by lovung** [comment]
@iapicca I think the 2nd solution is better. Because the targets of Flutter are not only on iOS and Android but also on other platforms.

**Comment by iapicca** [comment]
for my understanding [isar](https://github.com/isar/isar/tree/main/isar) from excellent @leisim
binds a rust base db using dart:ffi for both web and native
borrowing this approach could make the second option quite realistic

> 2. [QUICHE](https://github.com/cloudflare/quiche) + [dart:ffi](https://dart.dev/guides/libraries/c-interop) (resource for [rust + dart:ffi](https://medium.com/flutter-community/finally-running-rust-natively-on-a-flutter-plugin-here-is-how-6f2826eb1735)) could work from both android, ios, desktop and maybe in the future for [web](https://github.com/dart-lang/sdk/issues/37882)

**Comment by leisim** [comment]
@iapicca No, Isar will use the Rust engine for mobile and desktop targets and a custom typescript db engine for web. The reason is that WASM has no access to Browser APIs (IndexedDB) and has to communicate via JavaScript. Depending on the use case, this can be a lot slower than using JavaScript in the first place.

**Comment by iapicca** [comment]
@leisim 
thanks for clarifying that, 
I guess step 1 could be mobile only for now

**Comment by leisim** [comment]
@iapicca Unfortunately I don't think that would solve the performance problems. There are some outstanding [WASM proposals](https://github.com/WebAssembly/proposals) that will significantly improve the situation but they will need years until all browsers ship them.

**Comment by iapicca** [comment]
this [feature request](https://github.com/google/cronet.dart) could help help implementing the feature

**Comment by Sunbreak** [comment]
If anyone interested, [cronet_flutter](https://github.com/Sunbreak/cronet_flutter) is the origin PoC of [google/cronet.dart](https://github.com/google/cronet.dart/issues), and could run in `Android/iOS` and  `Windows/macOS/Linux`

**Comment by pomid** [comment]
i understand a native implementation might be difficult but a binding to quiche or ms-quic is not out of reach ...
where are we +2 years lates ?

**Comment by vukitoso** [comment]
Who is using what solutions now as an HTTP/3 server and client?

**Comment by Kellykinyama12** [comment]
We have started a pure dart dtls implementation on this repo:

- https://github.com/Kellykinyama12/dart-pure-dtls
The project can be repurposed to handle tls, as tls and dtls are similar. There is a lib\crypto.dart file that can handle cryptography issues

**Comment by KellyKinyama** [comment]
Quic implementation in pure dart:
https://github.com/KellyKinyama/pure-dart-quic

**Comment by KellyKinyama** [comment]
https://pub.dev/packages/pure_dart_quic

**Comment by iapicca** [comment]
> Quic implementation in pure dart: https://github.com/KellyKinyama/pure-dart-quic

https://raw.githubusercontent.com/KellyKinyama/pure-dart-quic/refs/heads/master/lib/pure_dart_quic.dart 😆

**Comment by KellyKinyama** [comment]
> > Quic implementation in pure dart: https://github.com/KellyKinyama/pure-dart-quic
> 
> https://raw.githubusercontent.com/KellyKinyama/pure-dart-quic/refs/heads/master/lib/pure_dart_quic.dart 😆


https://github.com/KellyKinyama/pure-dart-quic/blob/master/lib/connection/client/client3.dart

**Comment by KellyKinyama** [comment]
> > > Quic implementation in pure dart: https://github.com/KellyKinyama/pure-dart-quic
> > 
> > 
> > https://raw.githubusercontent.com/KellyKinyama/pure-dart-quic/refs/heads/master/lib/pure_dart_quic.dart 😆
> 
> https://github.com/KellyKinyama/pure-dart-quic/blob/master/lib/connection/client/client3.dart

https://github.com/KellyKinyama/pure-dart-quic/blob/master/lib/connection/server/server2.dart
